eclipse-theia/theia
# Refactor services parts of plugin-ext to use DI

#### What it does

This PR refactors the way services (CommandRegistryMain, QuickOpenMain, etc) are set up in RPCProtocol and makes all these services use DI.

This is part of https://github.com/eclipse-theia/theia/issues/5761 but also makes it easier to extend services.

#### How to test

Since this PR doesn't actually add any new features, it just has refactoring, we want to verify that all the services are working correctly.

This is a list of extensions you can install to test various parts of the plugin API:
Markdown Preview Enhanced -> ext install shd101wyy.markdown-preview-enhanced
Todo Tree -> ext install Gruntfuggly.todo-tree
JSON tree view -> ext install ChaunceyKiwi.json-tree-view
Git history -> ext install donjayamanne.githistory
https://github.com/JPinkney/Plugin-API-Test-vsix/blob/master/plugin-api-test-v2.vsix

This is a list of which parts of the plugin API are activated by which extensions
- [ ] - CommandRegistryMainImpl -> vscode-java 
- [ ] - QuickOpenMainImpl -> plugin-api-test-v2 -> From command palette run `Test quickopen`
- [ ] - LanguagesMainImpl -> vscode-java
- [ ] - OutputChannelRegistryMainImpl -> vscode-java with trace.server on
- [ ] - WorkspaceMainImpl -> Automatically used on start
- [ ] - DialogsMainImpl -> plugin-api-test-v2 -> From command palette run `Test dialog`
- [ ] - MessageRegistryMainImpl -> Install Git history and it should error and use doShowMessage from this class
- [ ] - PreferenceRegistryMainImpl -> vscode-java
- [ ] - StatusBarMessageRegistryMainImpl -> vscode-java
- [ ] - EnvMainImpl -> plugin-api-test-v2 -> From command palette run `Test env`
- [ ] - NotificationMainImpl -> vscode-java
- [ ] - TerminalServiceMainImpl
- [ ] - TreeViewsMainImpl -> JSON tree view -> open a json file
- [ ] - WebviewsMainImpl -> Install Markdown Preview Enhanced -> create file.md -> execute the `Open preview to the side` command
- [ ] - StorageMainImpl
- [ ] - ConnectionMainImpl
- [ ] - TasksMainImpl -> plugin-api-test-v2 -> From command palette run `Test task`
- [ ] - LanguagesContributionMainImpl
- [ ] - DebugMainImpl -> vscode java debug
- [ ] - ScmMainImpl -> Open a folder thats not a git repo -> open the source control view -> click the + at the top 
- [ ] - DecorationsMainImpl
- [ ] - WindowStateMain -> any, just unfocus then focus theia
- [ ] - ClipboardMainImpl -> plugin-api-test-v2 -> From command palette run `Test clipboard`
- [ ] - TextEditorsMainImpl -> automatically used
- [ ] - DocumentsMainImpl -> automatically used


#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
